File: 2023/13/13.rb
#!/usr/bin/env ruby
require 'pry'

MAPS = []

current_map = []
$stdin.readlines.lazy.map(&:chomp).each do |l|
  if l.empty?
    MAPS << current_map
    current_map = []
  else
    current_map << l.gsub('.', '0').gsub('#', '1').chars
  end
end

MAPS << current_map

def reflects?(current_map, row_index, diff = 0)
  first, second = current_map[0...row_index].reverse, current_map[row_index..]

  # Scale down to matching size
  if first.size < second.size
    second = second[0...first.size]
  else
    first = first[0...second.size]
  end

  diff_count = 0

  reflects = (0...first.size).all? do |i|
    next true if first[i] == second[i]
    values = [first[i], second[i]].sort
    return false unless values.first & values.last == values.first # Make sure bit difference doesn't trigger carry
    binary_diff = (values.last - values.first).to_s(2)
    count = binary_diff.chars.count { |c| c == '1' } # number of characters off
    if count + diff_count <= diff
      diff_count += count # keep track of how many differences
      true
    end
  end
  reflects and diff_count == diff
end

def binary_diff(a, b, diff = 1)
  a & b == a
end

def find_reflection(current_map, diff = 0)
  binary_map = current_map.map { |l| l.join('').to_i(2) }
  (1...binary_map.size).find do |idx|
    reflects?(binary_map, idx, diff)
  end || 0
end

sum = MAPS.each_with_index.map do |current_map, idx|
  row = find_reflection(current_map, 0)
  #puts "Found reflection on #{idx + 1} at row #{row}" if row > 0

  col = find_reflection(current_map.transpose, 0)
  #puts "Found reflection on #{idx + 1} at col #{col}" if col > 0
  row * 100 + col
end.sum
puts sum

sum = MAPS.each_with_index.map do |current_map, idx|
  row = find_reflection(current_map, 1)
  #puts "Found reflection on #{idx + 1} at row #{row}" if row > 0

  col = find_reflection(current_map.transpose, 1)
  #puts "Found reflection on #{idx + 1} at col #{col}" if col > 0
  row * 100 + col
end.sum
puts sum

File: 2023/13/13.go
package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"slices"
	"strconv"
)

func reflects(currentMap []uint, row_index int, diff int) bool {
	first, second := slices.Clone(currentMap[:row_index]), currentMap[row_index:]
	slices.Reverse(first)

	if len(first) < len(second) {
		second = second[:len(first)]
	} else if len(first) > len(second) {
		first = first[:len(second)]
	}

	diff_count := 0

	for i := range len(first) {
		var v1, v2 uint
		if first[i] == second[i] {
			continue
		} else if first[i] < second[i] {
			v1, v2 = first[i], second[i]
		} else {
			v1, v2 = second[i], first[i]
		}

		if v1&v2 != v1 {
			return false
		}
		count := bits.OnesCount(v2 - v1)

		if count+diff_count <= diff {
			diff_count += count
		} else {
			return false
		}
	}
	return diff_count == diff
}

func convertBinary(strMap []string) []uint {
	binaryMap := make([]uint, len(strMap))
	for i, v := range strMap {
		bv, _ := strconv.ParseInt(v, 2, 32)
		binaryMap[i] = uint(bv)
	}
	return binaryMap
}

func findReflection(currentMap []string, diff int) int {
	binaryMap := convertBinary(currentMap)
	for i := range len(binaryMap) - 1 {
		if reflects(binaryMap, i+1, diff) {
			return i + 1
		}
	}
	return 0
}

func transpose(strMap []string) []string {
	newMap := make([]string, len(strMap[0]))
	for i := 0; i < len(strMap[0]); i++ {
		row := make([]byte, len(strMap))
		for j := 0; j < len(strMap); j++ {
			row[j] = strMap[j][i]
		}
		newMap[i] = string(row)
	}
	return newMap
}

func sumReflections(maps [][]string, diff int) int {
	sum := 0
	for _, currentMap := range maps {
		sum += 100 * findReflection(currentMap, diff)
		sum += findReflection(transpose(currentMap), diff)
	}
	return sum
}

func ParseInput() [][]string {
	scanner := bufio.NewScanner(os.Stdin)
	maps := [][]string{}
	currentMap := []string{}

	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			maps = append(maps, currentMap)
			currentMap = []string{}
			continue
		}
		row := ""
		for _, c := range line {
			if c == '#' {
				row += "1" // uint(math.Pow(2, float64(i)))
			} else {
				row += "0"
			}
		}
		currentMap = append(currentMap, row)
	}
	return append(maps, currentMap)
}

func main() {
	maps := ParseInput()
	fmt.Println(sumReflections(maps, 0))
	fmt.Println(sumReflections(maps, 1))
}
